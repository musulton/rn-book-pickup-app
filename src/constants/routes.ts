const ROUTES = {
    HOME: "home",
    BOOKS: "books",
    BOOK_DETAILS: "bookDetails",
    HISTORY: "history"
}

export default ROUTES
