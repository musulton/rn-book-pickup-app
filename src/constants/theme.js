import { Dimensions } from "react-native";
const { width, height } = Dimensions.get("window");

export const COLORS = {
    // base colors
    primary: "#039BE5",
    secondary: "#25282F",

    // colors
    black: "#1E1B26",
    white: "#FFFFFF",
    lightGray: "#64676D",
    lightGray2: "#EFEFF0",
    lightGray3: '#D4D5D6',
    lightGray4: '#7D7E84',
    gray: "#2D3038",
    gray1: "#282C35",
    darkRed: "#31262F",
    lightRed: "#C5505E",
    darkBlue: "#22273B",
    lightBlue: "#424BAF",
    darkGreen: "#213432",
    lightGreen: "#31Ad66",

};

export const SIZES = {
    padding1: 20,
    padding2: 10,
    borderRadius: 10,

    // app dimensions
    width,
    height
};

const appTheme = { COLORS, SIZES };

export default appTheme;
