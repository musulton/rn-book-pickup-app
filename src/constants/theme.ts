import { Dimensions } from "react-native";
const { width, height } = Dimensions.get("window");

export const COLORS = {
    primary: "#039BE5",
    background: "#FEFEFE",

    black: "#1E1B26",
    white: "#FFFFFF",
    lightGray: "#64676D",
    lightGray2: "#EFEFF0",
    lightGray3: '#D4D5D6',
};

export const SIZES = {
    padding1: 20,
    padding2: 10,
    borderRadius: 6,

    // app dimensions
    width,
    height
};

const appTheme = { COLORS, SIZES };

export default appTheme;
