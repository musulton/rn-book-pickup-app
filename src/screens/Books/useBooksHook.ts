import React from "react";
import useFetch from "../../hooks/useFetch";
import {getBooks} from "../../services/api";
import {Book} from "../../types";
import {getDummyCoverImg} from "../../utils/function";

type CurrentPageFn = (prevState: number) => number
interface UseBookEffect {
    getter: {
        data: Array<Book>
        error: boolean
        loading: boolean
    }
    setter: {
        currentPage: (page: number | CurrentPageFn) => void
    }
}

const useBookEffect: () => UseBookEffect = () => {
    const [currentPage, setCurrentPage] = React.useState(1)
    const [bookList, setBookList] = React.useState<Array<Book>>([])
    const {data, error, loading} = useFetch(getBooks, { page: currentPage })

    React.useEffect(() => {
        const dataTransform: Array<Book> = data?.data?.works?.map((book: any) => ({
            title: book?.title,
            author: book?.authors[0]?.name,
            edition: book?.["cover_edition_key"],
            coverImg: getDummyCoverImg()
        })) || []

        const newBooks: Array<Book> = [...bookList, ...dataTransform]
        setBookList(newBooks as Array<Book>)
    }, [data?.data?.works])

    return {
        getter: { data: bookList, error, loading },
        setter: { currentPage: setCurrentPage }
    }
}

export default useBookEffect
