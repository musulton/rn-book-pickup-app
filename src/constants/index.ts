import theme, { COLORS, SIZES } from "./theme";
import ROUTES from "./routes";

export { theme, COLORS, SIZES, ROUTES };
