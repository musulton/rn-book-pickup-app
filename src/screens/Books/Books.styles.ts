import {StyleSheet, ViewStyle} from "react-native";

import {COLORS, SIZES} from "../../constants";

interface Styles {
    container: ViewStyle
    emptyState: ViewStyle
    scrollContainer: ViewStyle
}

const styles = StyleSheet.create<Styles>({
    container: {
      backgroundColor: COLORS.background,
        paddingTop: SIZES.padding1
    },
    scrollContainer: {
        display: "flex",
        flex: 1,
        backgroundColor: COLORS.background,
    },
    emptyState: {
        display: "flex",
        flex: 1,
        justifyContent: "center",
        alignItems: "center"
    }
});

export default styles
