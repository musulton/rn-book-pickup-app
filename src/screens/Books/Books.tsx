import {ActivityIndicator, FlatList, ListRenderItemInfo, SafeAreaView} from "react-native";
import React from "react";
import {NavigationProp} from "@react-navigation/native";

import Header from "../../components/Header";
import EmptyState from "../../components/EmptyState";
import CardBook from "../../components/CardBook";
import {ROUTES} from "../../constants";
import {Book} from "../../types";

import styles from "./Books.styles";
import useBooksHook from "./useBooksHook";

interface BooksProps {
    navigation: NavigationProp<any>
}

const Books: React.FC<BooksProps> = (props) => {
    const { getter, setter } = useBooksHook();
    const hasData = getter?.data?.length > 0

    const onNavigate = React.useCallback((book: Book) => () => {
        props.navigation.navigate(ROUTES.BOOK_DETAILS, {
            book
        })
    }, [])

    const onEndReached = () => {
        setter.currentPage((currentPage: number) => currentPage + 1)
    }

    return (
        <SafeAreaView style={styles.container}>
            <Header />
            <FlatList
                data={getter?.data}
                renderItem={(props: ListRenderItemInfo<Book>) => (
                    <CardBook
                        title={props.item.title}
                        author={props.item.author}
                        edition={props.item.edition}
                        coverImg={props.item.coverImg}
                        onPress={onNavigate(props.item)}
                    />
                )}
                onEndReached={onEndReached}
                refreshing={getter.loading}
                contentContainerStyle={!hasData && styles.scrollContainer}
                ListEmptyComponent={<EmptyState />}
                ListFooterComponent={hasData ? <ActivityIndicator size={"small"} /> : null}
                initialNumToRender={10}
                testID={"bookList"}
            />
        </SafeAreaView>
    )
}

export default Books
