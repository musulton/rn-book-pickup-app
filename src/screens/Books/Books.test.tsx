import renderer from "react-test-renderer";
import {fireEvent, render, screen} from "@testing-library/react-native";

import {ROUTES} from "../../constants";
import {book} from "../../fixtures/testData";

import Books from "./Books";
import useBooksHook from "./useBooksHook";

jest
    .mock("./useBooksHook")
    .mock("../../components/EmptyState", () => "EmptyState")
    .mock("../../components/Header", () => "Header")

describe("<Books />", () => {
    it("should render Books correctly", () => {
        useBooksHook.mockReturnValue({
            getter: {
                data: [book]
            }
        })

        render(<Books />)

        expect(screen.toJSON()).toMatchSnapshot()
    })

    it("should render History correctly when empty state", () => {
        useBooksHook.mockReturnValue({
            getter: {
                data: []
            }
        })

        render(<Books />)

        expect(screen.toJSON()).toMatchSnapshot()
    })

    it("should navigate to <BookDetails /> when <CardBook /> is pressed", async () => {
        const mockProps = {
            navigation: {
                navigate: jest.fn()
            }
        }

        useBooksHook.mockReturnValue({
            getter: {
                data: [book]
            }
        })

        render(<Books {...mockProps} />)

        const cardItem = await screen.findByText("title")
        fireEvent.press(cardItem)

        expect(mockProps.navigation.navigate).toHaveBeenCalledWith(ROUTES.BOOK_DETAILS, {
            book: book
        })
    })

    it("should call set current page when scroll to the end of the list", async () => {
        const mockSetCurrentPage = jest.fn()

        useBooksHook.mockReturnValue({
            getter: {
                data: [book]
            },
            setter: {
                currentPage: mockSetCurrentPage
            }
        })

        render(<Books />)

        const bookList = await screen.findByTestId("bookList")

        fireEvent.scroll(bookList, {
            nativeEvent: {
                contentOffset: {
                    y: 500,
                },
                contentSize: {
                    height: 500,
                    width: 100,
                },
                layoutMeasurement: {
                    height: 100,
                    width: 100,
                },
            }
        })

        expect(mockSetCurrentPage).toHaveBeenCalledWith(expect.any(Function))
    })
})
