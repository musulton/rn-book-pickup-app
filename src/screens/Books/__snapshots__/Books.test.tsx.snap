// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Books /> should render Books correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#FEFEFE",
      "paddingTop": 20,
    }
  }
>
  <Header />
  <RCTScrollView
    ListEmptyComponent={<EmptyState />}
    ListFooterComponent={
      <ActivityIndicator
        size="small"
      />
    }
    contentContainerStyle={false}
    data={
      [
        {
          "author": "author",
          "edition": "edition",
          "title": "title",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    initialNumToRender={10}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="bookList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "marginVertical": 10,
              "opacity": 1,
              "paddingHorizontal": 10,
            }
          }
        >
          <View
            style={
              {
                "marginRight": 5,
                "resizeMode": "stretch",
              }
            }
          >
            <Image
              source={
                {
                  "uri": undefined,
                }
              }
              style={
                {
                  "borderRadius": 6,
                  "height": 100,
                  "resizeMode": "cover",
                  "width": 70,
                }
              }
            />
          </View>
          <View
            style={
              {
                "alignContent": "center",
                "display": "flex",
                "flex": 1,
                "justifyContent": "center",
                "marginLeft": 10,
              }
            }
          >
            <Text
              numberOfLines={2}
              style={
                {
                  "fontSize": 18,
                  "fontWeight": "500",
                }
              }
            >
              title
            </Text>
            <Text
              numberOfLines={1}
              style={
                {
                  "color": "#64676D",
                  "marginVertical": 5,
                }
              }
            >
              author
            </Text>
            <Text
              numberOfLines={1}
              style={
                {
                  "fontSize": 16,
                  "fontWeight": "500",
                }
              }
            >
              edition
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <ActivityIndicator
          size="small"
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`<Books /> should render History correctly when empty state 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#FEFEFE",
      "paddingTop": 20,
    }
  }
>
  <Header />
  <RCTScrollView
    ListEmptyComponent={<EmptyState />}
    ListFooterComponent={null}
    contentContainerStyle={
      {
        "backgroundColor": "#FEFEFE",
        "display": "flex",
        "flex": 1,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    initialNumToRender={10}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="bookList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <EmptyState
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
